### PR TITLE
Ft task#6 create card

### PR DESCRIPTION
**Pull request by akrawczyk2021**
I`ve lost my Integration tests i`ll do it in separate task.

AC
Create Card with validation
Unit tests
API tests

### CODE REVIEW

File: src/Repository/CardRepository.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        $this->entityManager->persist($card);
+    }
+
+    public function findOneByName(string $name): ?Card
+    {
+        return $this->entityManager->getRepository(Card::class)->findOneBy(['name' => $name]);
+    }
+
+    public function createCard(CreateCard $createCard): void
```
Comment (jlackipl) [inline src/Repository/CardRepository.php]: This is not correct place to convert DTO into Entity. In basic you put responsibility of creating Entity object in repository but this responsibility is not related to this class.

Thing about it. How many reasons for change we currently have in repository? 
1. One is when something change related to fetching entities from DB.
2. Next one is when we change something what is related to Create Entity for example we add new property.

So if we have more then one reason to change there is a chance that we do something wrong. Here IMO we brake SRP of repository. In some cases we don't follow that rule but we need to good reason for it.

We have multiple ways to make it better.:
1. Put it into controller. Why? Cuz this logic is simple and if we don't better solution then make it simply and strait forward. And from there you could look on it and thing once more how to refactor it.
2. Create handler like`CreateCardHandler::handle(CreateCard): void`, that handle Domain Validation, Converting DTO to Entity and invoke `add` method on repository.

I suggest option one for now. If we gather more examples then we make a call and discuss it.

File: tests/acceptance.suite.yml [hunk 2]
```diff
@@ -1,12 +1,4 @@
...
-# Perform tests in browser using the WebDriver or PhpBrowser.
-# If you need both WebDriver and PHPBrowser tests - create a separate suite.
-
 actor: AcceptanceTester
 modules:
     enabled:
-        - PhpBrowser:
-            url: '192.168.0.133:80'
-        
+        - \App\Tests\Helper\Acceptance
```
Comment (jlackipl) [inline tests/acceptance.suite.yml]: Add empty line at the end of file

File: tests/api.suite.yml [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
+        url: /
+        depends: Symfony
+    - Db:
+        dsn: 'mysql:host=mysql;dbname=symfonyak'
+        user: 'root'
+        password: 'haslo'
+        populate: true
+        populator: 'php bin/console d:m:m --no-interaction --allow-no-migration'
+        cleanup: true
+    - App\Tests\Helper\Api
```
Comment (jlackipl) [inline tests/api.suite.yml]: Add empty line at EOF

File: tests/acceptance/CreateCardCest.php [hunk 4]
```diff
@@ -1,19 +1,13 @@
 <?php
 namespace App\Tests;
 use App\Tests\AcceptanceTester;
-class FirstCest
+class CreateCardCest
```
Comment (jlackipl) [inline tests/acceptance/CreateCardCest.php]: Remove Acceptance folder and any configuration for it

File: tests/api/CreateCardEndpointCest.php [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                'name' => 'Goblin',
+                'description' => 'Low level monster',
+                'hp' => 10,
+                'attack' => 1,
+                'defense' => 1,
+            ]
+        );
+    }
+
+    public function testItThrowBadRequestWithNotUniqueName(ApiTester $I)
```
Comment (jlackipl) [inline tests/api/CreateCardEndpointCest.php]: Try to use data provider for BadRequest tests

File: tests/unit/CardDataValidatorTest.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                'attack' => 0
+            ],
+        ];
+    }
+
+    /**
+     * @param int $value
+     * @dataProvider incorrectStatValuesDataProvider
+     */
+    public function testIsIntStatInvalidWithInorrectData(int $value)
```
Comment (jlackipl) [inline tests/unit/CardDataValidatorTest.php]: typo

File: tests/unit/CardDataValidatorTest.php [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        $this->assertFalse($this->validator->isValidIntStat($value));
+    }
+
+    public function incorrectStatValuesDataProvider()
+    {
+        return [
+            'nagative value' => [
+                'attack' => -5
+            ],
+
```
Comment (jlackipl) [inline tests/unit/CardDataValidatorTest.php]: rm line

File: tests/unit/CardDataValidatorTest.php [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+use App\Repository\CardRepository;
+use App\Validator\CardDataValidator;
+use Exception;
+
+class CardDataValidatorTest extends \Codeception\Test\Unit
+{
+    /**
+     * @var \App\Tests\UnitTester
+     */
+    protected $tester;
```
Comment (jlackipl) [inline tests/unit/CardDataValidatorTest.php]: type hints

File: src/Validator/CardDataValidator.php [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+    public function isValidIntStat(int $stat): bool
+    {
+        return $stat >= 0;
+    }
+
+    public function ensureNameIsUnique(string $name): void
+    {
+        if ($this->cardRepository->findOneByName($name)) {
+            throw new BadRequestException("Name is already in use");
```
Comment (jlackipl) [inline src/Validator/CardDataValidator.php]: add test for it

File: src/Validator/CardDataValidator.php [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    public function isCardDataValid(string $name, string $description, int $attack, int $defence, int $hp): bool
```
Comment (jlackipl) [inline src/Validator/CardDataValidator.php]: this method should be on top

File: src/Validator/CardDataValidator.php [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+declare(strict_types=1);
+
+namespace App\Validator;
+
+use App\Repository\CardRepository;
+use Symfony\Component\HttpFoundation\Exception\BadRequestException;
+
+class CardDataValidator
+{
+    private $cardRepository;
```
Comment (jlackipl) [inline src/Validator/CardDataValidator.php]: typehint

Comment (jlackipl) [reply src/Validator/CardDataValidator.php]: try to use php 8 constructor argument promotion :)

File: src/ParamConverter/CreateCardParamConverter.php [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        $this->validator = $validator;
+        $this->cardRepository = $repository;
+    }
+
+    public function apply(Request $request, ParamConverter $configuration)
+    {
+        $content = json_decode($request->getContent(), true, 512, JSON_THROW_ON_ERROR);
+
+        if (
+            $this->validator->isCardDataValid(
```
Comment (jlackipl) [inline src/ParamConverter/CreateCardParamConverter.php]: I suggest something like:
```
$cardDTO = new CreateCard(
     $this->getName($requestContent),
     ... // and so on
);
```
So for each of property create a method that extract value from jsonContent. And also i that methods perform validation. So you could make it more readable. Also after that you could get rid of `isCardDataValid()`
And thanks to that you could set BadRequest message for each error so user get more explicite information whats wrong.

File: src/Validator/CardDataValidator.php (hunk 11 above)
Comment (akrawczyk2021) [reply src/Validator/CardDataValidator.php]: done

File: src/Validator/CardDataValidator.php (hunk 10 above)
Comment (akrawczyk2021) [reply src/Validator/CardDataValidator.php]: done

File: tests/unit/CardDataValidatorTest.php (hunk 8 above)
Comment (akrawczyk2021) [reply tests/unit/CardDataValidatorTest.php]: done

File: tests/unit/CardDataValidatorTest.php (hunk 7 above)
Comment (akrawczyk2021) [reply tests/unit/CardDataValidatorTest.php]: done

File: tests/unit/CardDataValidatorTest.php (hunk 6 above)
Comment (akrawczyk2021) [reply tests/unit/CardDataValidatorTest.php]: done

File: tests/acceptance/CreateCardCest.php (hunk 4 above)
Comment (akrawczyk2021) [reply tests/acceptance/CreateCardCest.php]: done

File: tests/api.suite.yml (hunk 3 above)
Comment (akrawczyk2021) [reply tests/api.suite.yml]: done

File: tests/acceptance.suite.yml (hunk 2 above)
Comment (akrawczyk2021) [reply tests/acceptance.suite.yml]: done

File: src/ParamConverter/CreateCardParamConverter.php (hunk 12 above)
Comment (akrawczyk2021) [reply src/ParamConverter/CreateCardParamConverter.php]: done please verify

File: src/ParamConverter/CreateCardParamConverter.php [hunk 13]
```diff
@@ -14,38 +14,23 @@
 
 class CreateCardParamConverter implements ParamConverterInterface
 {
-    private $validator;
+    private mixed $content;
```
Comment (jlackipl) [inline src/ParamConverter/CreateCardParamConverter.php]: Not mixed, actually it should be array

**Review by jlackipl — DISMISSED** [review]

File: src/Validator/CardDataValidator.php (hunk 9 above)
Comment (akrawczyk2021) [reply src/Validator/CardDataValidator.php]: done

File: src/ParamConverter/CreateCardParamConverter.php (hunk 13 above)
Comment (akrawczyk2021) [reply src/ParamConverter/CreateCardParamConverter.php]: done

File: tests/api/CreateCardEndpointCest.php (hunk 5 above)
Comment (akrawczyk2021) [reply tests/api/CreateCardEndpointCest.php]: done

File: src/Repository/CardRepository.php (hunk 1 above)
Comment (akrawczyk2021) [reply src/Repository/CardRepository.php]: done

**Review by jlackipl — APPROVED** [review]

**Pull request merged**